core/primitives/src/types.rs

```diff
@@ -422,39 +669,197 @@ pub struct BlockExtra {
     pub challenges_result: ChallengesResult,
 }
 
-/// Information after chunk was processed, used to produce or check next chunk.
-#[derive(Debug, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Clone, Eq)]
-pub struct ChunkExtra {
-    /// Post state root after applying give chunk.
-    pub state_root: StateRoot,
-    /// Root of merklizing results of receipts (transactions) execution.
-    pub outcome_root: CryptoHash,
-    /// Validator proposals produced by given chunk.
-    pub validator_proposals: Vec<ValidatorStake>,
-    /// Actually how much gas were used.
-    pub gas_used: Gas,
-    /// Gas limit, allows to increase or decrease limit based on expected time vs real time for computing the chunk.
-    pub gas_limit: Gas,
-    /// Total balance burnt after processing the current chunk.
-    pub balance_burnt: Balance,
-}
-
-impl ChunkExtra {
-    pub fn new(
-        state_root: &StateRoot,
-        outcome_root: CryptoHash,
-        validator_proposals: Vec<ValidatorStake>,
-        gas_used: Gas,
-        gas_limit: Gas,
-        balance_burnt: Balance,
-    ) -> Self {
-        Self {
-            state_root: state_root.clone(),
-            outcome_root,
-            validator_proposals,
-            gas_used,
-            gas_limit,
-            balance_burnt,
+#[cfg(feature = "protocol_feature_block_header_v3")]
+pub mod chunk_extra {
+    use crate::types::validator_stake::{ValidatorStake, ValidatorStakeIter, ValidatorStakeV1};
+    use crate::types::StateRoot;
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use near_primitives_core::hash::CryptoHash;
+    use near_primitives_core::types::{Balance, Gas};
+    use serde::Serialize;
+
+    /// Information after chunk was processed, used to produce or check next chunk.
+    #[derive(Debug, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Clone, Eq)]
+    pub enum ChunkExtra {
+        V1(ChunkExtraV1),
+        V2(ChunkExtraV2),
+    }
```

Comment:
I removed `Serialize` from most structs in https://github.com/near/nearcore/pull/4115/commits/f9be8cb1cf5adfa198f86d8df78b5a78d335d51b

`ValidatorStake` was left because its serialization is used (at least removing `Serialize` made the compiler unhappy). But I added the tagged attribute same as in `ValidatorStakeView`.